File: ql/src/experimental/Security/CWE-643/XPathInjection.go
package main

import (
  "net/http"
  "github.com/moovweb/gokogiri"
)

func processRequest(r *http.Request, doc *XmlDocument) {
  r.parseForm()
  username := r.Form.Get("username")
  password := r.Form.Get("password")
  
  root := doc.Root()
  // BAD: User input used directly in an XPath expression
  doc, _ := root.SearchWithVariables("//users/user[login/text()='" + username + "' and password/text() = '" + password + "']/home_dir/text()")

  // GOOD: Uses parameters to avoid including user input directly in XPath expression
  doc, _ := root.SearchWithVariables("//users/user[login/text()=$username and password/text() = $password]/home_dir/text()")
}

File: csharp/ql/src/Security Features/CWE-643/XPathInjection.cs
using System;
using System.Web;
using System.Xml.XPath;

public class XPathInjectionHandler : IHttpHandler
{
    public void ProcessRequest(HttpContext ctx)
    {
        string userName = ctx.Request.QueryString["userName"];

        // BAD: Use user-provided data directly in an XPath expression
        string badXPathExpr = "//users/user[login/text()='" + userName + "']/home_dir/text()";
        XPathExpression.Compile(badXPathExpr);

        // GOOD: XPath expression uses variables to refer to parameters
        string xpathExpression = "//users/user[login/text()=$username]/home_dir/text()";
        XPathExpression xpath = XPathExpression.Compile(xpathExpression);

        // Arguments are provided as a XsltArgumentList()
        XsltArgumentList varList = new XsltArgumentList();
        varList.AddParam("userName", string.Empty, userName);

        // CustomContext is an application specific class, that looks up variables in the
        // expression from the varList.
        CustomContext context = new CustomContext(new NameTable(), varList)
      xpath.SetContext(context);
    }
}
